### PR TITLE
build: optimize docker build by better utilizing build cache

### PR DESCRIPTION
**Pull request by pwang2**
### SUMMARY
Using common docker build practice to use docker build cache better.  
New docker image build
1. 394.79 MB -> 348.48 MB
2. 3-5 minutes saving for initial build by not copying/deleting unneeded file, especial node_modules.
3. merged the first stage to the lean image. still being able to cache the pip requirements installation. 
4. fix the random error when delete node_modules folder.
image

### TESTING INSTRUCTIONS

```bash
podman build -t superset  --no-cache --target lean -f ./Dockerfile
```

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by EugeneTorap** [comment]
@pwang2 Thank you so much for this PR!
The changes should speeds up docker build CI by 3-5 minutes.

File: Dockerfile [hunk 1]
```diff
@@ -15,60 +15,34 @@
...
-        && npm ci
+RUN /frontend-mem-nag.sh
 
-# This seems to be the most expensive step
-RUN cd /app/superset-frontend \
-        && npm run ${BUILD_CMD} \
-        && rm -rf node_modules
+WORKDIR /app/superset-frontend/
 
+COPY superset-frontend/package*.json .
```
Comment (EugeneTorap) [inline Dockerfile:35]: **When using COPY with more than one source file, the destination must be a directory and end with a /**
```suggestion
COPY superset-frontend/package*.json ./
```

**Comment by pwang2** [comment]
Receiving a lot of GitHub Actions failure which I believe should nothing to do with the change here.   Can anyone help me figure out if anything missing from my side.

File: Dockerfile [hunk 2]
```diff
@@ -15,60 +15,34 @@
...
-RUN cd /app \
-    && mkdir -p superset/static \
-    && touch superset/static/version_info.json \
-    && pip install --no-cache -r requirements/local.txt
-
-
 ######################################################################
 # Node stage to deal with static asset construction
 ######################################################################
+ARG PY_VER=3.8.13-slim
```
Comment (zhaoyongjie) [inline Dockerfile:21]: The section is the Node image so I think `PY_VER` is useless here.

File: Dockerfile [hunk 3]
```diff
@@ -85,16 +59,24 @@ RUN mkdir -p ${PYTHONPATH} \
...
         && rm -rf /var/lib/apt/lists/*
 
-COPY --from=superset-py /usr/local/lib/python3.8/site-packages/ /usr/local/lib/python3.8/site-packages/
-# Copying site-packages doesn't move the CLIs, so let's copy them one by one
-COPY --from=superset-py /usr/local/bin/gunicorn /usr/local/bin/celery /usr/local/bin/flask /usr/bin/
+COPY ./requirements/*.txt  /app/requirements/
+COPY setup.py MANIFEST.in README.md /app/
+
+# setup.py uses the version information in package.json
+COPY superset-frontend/package.json /app/superset-frontend/
```
Comment (zhaoyongjie) [inline Dockerfile:72]: IMO setup.py to read package.json isn't a good idea but exists a long time in Superset, might change this logic in the separated PR.

**Review by zhaoyongjie — APPROVED** [review]
LGTM, leave one question for the `py_ver` argument in Node section.

File: Dockerfile (hunk 2 above)
Comment (pwang2) [reply Dockerfile:21]: This is some not well documented in docker document.  in a multistage dockerfile, ARG used in FROM need to be added before the first FROM. 

image

**Pull request merged**